src/hiring.rs

```diff
@@ -0,0 +1,285 @@
+
+        // Set of identifiers for all applications which have been added, but not removed, for this opening.
+        // Is required for timely on-chain lookup of all applications associated with an opening.
+        applicants: BTreeSet<ApplicationId>, //BTreeMap<ApplicationId, ()>, //Vec<T::ApplicationId>,
+
+        // TODO: Drop these counters
+        // https://github.com/Joystream/substrate-hiring-module/issues/9
+        //
+        // Counters over all possible application states.
+        // Are needed to set `application_index_in_opening` in new applications
+        // Are very useful for light clients.
+        //
+        // NB: Remember that _all_ state transitions in applications will require updating these variables,
+        // its easy to forget!
+        //
+        // NB: The sum of
+        // - `active_application_count`
+        // - `unstaking_application_count`
+        // - `deactivated_application_count`
+        //
+        // equals the total number of applications ever added to the openig via `add_application`.
+
+        // Active NOW
+        active_application_count: u32,
+
+        // Unstaking NOW
+        unstaking_application_count: u32,
+
+        // Deactivated at any time for any cause.
+        deactivated_application_count: u32, // Removed at any time.
+                                            //removed_application_count: u32
+    },
+}
+
+impl<BlockNumber, ApplicationId> OpeningStage<BlockNumber, ApplicationId> {
+    /// The number of applications ever added to the opening via
+    /// `add_opening` extrinsic.
+    pub fn number_of_appliations_ever_added(&self) -> u32 {
+        match self {
+            OpeningStage::WaitingToBegin { .. } => 0,
+
+            OpeningStage::Active {
+                active_application_count,
+                unstaking_application_count,
+                deactivated_application_count,
+                ..
+            } => {
+                active_application_count
+                    + unstaking_application_count
+                    + deactivated_application_count
+            }
+        }
+    }
+}
+
+/// OpeningStage must be default constructible because it indirectly is a value in a storage map.
+/// ***SHOULD NEVER ACTUALLY GET CALLED, IS REQUIRED TO DUE BAD STORAGE MODEL IN SUBSTRATE***
+impl<BlockNumber: Default, ApplicationId> Default for OpeningStage<BlockNumber, ApplicationId> {
+    fn default() -> Self {
+        OpeningStage::WaitingToBegin {
+            begins_at_block: BlockNumber::default(),
+        }
+    }
+}
+
+// Constraints around staking amount
+#[derive(Encode, Decode, Debug, Eq, PartialEq, Clone)]
+pub enum StakingAmountLimitMode {
+    AtLeast,
+    Exact,
+}
+
+// Policy for staking
+#[derive(Encode, Decode, Debug, Eq, PartialEq, Clone)]
+pub struct StakingPolicy<Balance, BlockNumber> {
+    // Staking amount
+    pub amount: Balance,
+
+    // How to interpret the amount requirement
+    pub amount_mode: StakingAmountLimitMode,
```

Comment:
I don't see a strong case for it, we would not eliminate any bad states, and I don't think we avoid any repetition. Given this would break downstream dependencies, and some time to just work internally, it does not seem worth it.

Reply:
agree